### PR TITLE
On-type formatter improvements

### PR DESCRIPTION
**Pull request by tomaz**
Previously, the text was always trimmed at end, but this interfered with preformatted text when on-type formatter was used. With this change, typing is much more fluent.

Also added enter key as on-type format trigger. When pressed, format for previous line is invoked. This works quite well when typing but it invokes formatter also for example when initially opening the file and pressing enter to start typing new line. It should be fine given same formatting is usually desired over the whole file, but may be unexpected. Not sure if this is detectable though...

I tested the change for paste and format on save too, as well as for different formatting options, and all was fine. But am not very knowledgable in JS, so maybe I missed something. If anything raises your concerns, please let me know!

### CODE REVIEW

**Comment by tomaz** [comment]
Oh, just noticed, an extra space is added between arguments if enter is pressed (for `LD A, B` for example). Works for `LD A, <num>`. Will investigate and update...

**Comment by tomaz** [comment]
Ok, got it, extra space is only inserted if pressing enter for one of intellisense suggestions. In this case formatter is not invoked. Personally, I can live with this, but will try to find a solution nonetheless (any hint would be appreciated anyway 😉)

File: src/formatProcessor.ts
Comment (mborik) [inline src/formatProcessor.ts]: What is purpose of `char`?
Therefore that you didn't use it anywhere I'm suggesting following simplifications:

File: src/formatProcessor.ts
Comment (mborik) [inline src/formatProcessor.ts]:
```suggestion
	format(document: vscode.TextDocument, range: vscode.Range, isOnType?: boolean): FormatProcessorOutput {
```

File: src/formatProcessor.ts
Comment (mborik) [inline src/formatProcessor.ts]: `var` is restricted by eslint, please follow that rules and turn on the linter in VSCode.
```suggestion
			let result = newText.join('');
```

File: src/formatProcessor.ts
Comment (mborik) [inline src/formatProcessor.ts]:
```suggestion
			if (!isOnType) {
```

File: src/formatProcessor.ts
Comment (mborik) [inline src/formatProcessor.ts]:
```suggestion
		return this.formatter.format(document, document.lineAt(line).range, true);
```

**Review by mborik — CHANGES_REQUESTED** [review]
Wow! 😌 
Thank you for such improvements. Personally I didn't use any kind of auto-formatters on type/save and I've couldn't investigate all that nuances of extension's behaviour of this feature. So I'm very thankful for that and I have just few minor objections: 👍

File: src/formatProcessor.ts
Comment (tomaz) [reply src/formatProcessor.ts]: I thought I would handle it inside formatter, but later realized I don't have info on line in the function. Agree it makes no sense to keep it in current solution, just adds the noise...

File: src/formatProcessor.ts
Comment (tomaz) [reply src/formatProcessor.ts]: Nice! Wasn't aware of this, long time programmer in many languages, but 0 experience with JS 😅 Enabled eslint and of course, it immediately pointed out this line 👍

**Comment by tomaz** [comment]
Addressed all comments + fixed another `var` instance at the bottom of the file + converted `==` to `===`, both pointed out by eslint. Let me know if you have any further comments! I'll squash correction commit(s) once approved.

Regardless, even if all ok, don't merge yet, I want to check the mentioned issue with intellisense as well. I'll let you know if I don't find a solution...

**Review by mborik — APPROVED** [review]

**Comment by mborik** [comment]
Great then, I will not merge it yet. I'm looking forward for another useful improvements from you. 👍
_(Manual squashing is not needed, github will squash it for us, I'll just click on "Squash and merge".)_

**Comment by tomaz** [comment]
github as the app? Don't see this option in web interface... Though I'm used to git, use squash and other commands daily.

**Comment by mborik** [comment]
> Don't see this option in web interface...

It's visible to repo owner. 😉

Manual squashing and rebasing in review process is bit tricky, all comments will lost it's track.

**Comment by mborik** [comment]
> an extra space is added between arguments if enter is pressed

It should be affected by `z80-macroasm.format.spaceAfterArgument` configuration parameter. If you have set it to `true`, it will add an extra space after comma.

**Comment by tomaz** [comment]
That option is fine by itself, it's `Z80CompletionProposer` that also inserts a space, so in some cases, 2 spaces are inserted - first by `Z80TypingFormatter`, then by completion proposer. I know exactly which lines are responsible, so my idea is to check if `formatOnType` is enabled, and if so, not insert the space (since it was already inserted by completion proposer).

What I'm struggling right now is to determine the value of `formatOnType`. There are 2 possibilities: either it's enabled for `z80-macroasm` file type or globally for every file. I can see the config values in debugger (`vscode.workspace.getConfiguration()`), just not able to read the value from it - mainly due to being unfamiliar with JS I think....

**Comment by mborik** [comment]
Don't be afraid, it's pretty easy. Just extend the `ConfigProps` interface in `configProperties.ts` with `formatOnType` prop and then fill it up by `config.editor.formatOnType` value in `getConfigProps()` method.
After that you will have formatOnType available in every provider. 😉

**Comment by tomaz** [comment]
oh wow, that's much simpler than I thought... I assumed that adding anything to `ConfigProps` would also affect extension settings... though `editor.formatOnType` is `false` in my case because I only have this enabled for `[z80-macroasm]`

image

**Comment by tomaz** [comment]
gotcha - `config.get("[z80-macroasm]")`!!!

**Comment by mborik** [comment]
I've got it. We need to specify the scope for 'getConfiguration`:
```diff
diff --git a/src/configProperties.ts b/src/configProperties.ts
index 1842fb8..082738a 100644
--- a/src/configProperties.ts
+++ b/src/configProperties.ts
@@ -1,4 +1,5 @@
 import * as vscode from 'vscode';
+import { EXTENSION_LANGUAGE_ID } from './extension';
 
 export interface ConfigProps {
 	indentSpaces: boolean;
@@ -21,7 +22,7 @@ export abstract class ConfigPropsProvider {
 	constructor(public settings: vscode.WorkspaceConfiguration) {}
 
 	getConfigProps(document: vscode.TextDocument) {
-		const config = vscode.workspace.getConfiguration();
+		const config = vscode.workspace.getConfiguration(undefined, { languageId: EXTENSION_LANGUAGE_ID });
 
 		const result: ConfigProps = {
 			...this.settings?.format,
diff --git a/src/extension.ts b/src/extension.ts
index 246af2b..888fa61 100644
--- a/src/extension.ts
+++ b/src/extension.ts
@@ -8,6 +8,7 @@ import { Z80HoverProvider } from './hover';
 import { Z80RenameProvider } from './renameProvider';
 import { Z80DocumentSymbolProvider, Z80WorkspaceSymbolProvider } from './symbolProvider';
 
+export const EXTENSION_LANGUAGE_ID = 'z80-macroasm';
 
 let changeConfigSubscription: vscode.Disposable | undefined;
 let symbolProcessor: SymbolProcessor | undefined;
@@ -34,13 +35,12 @@ export function deactivate() {
 }
 
 function configure(ctx: vscode.ExtensionContext, event?: vscode.ConfigurationChangeEvent) {
-	const language = 'z80-macroasm';
-	const settings = vscode.workspace.getConfiguration(language);
-	const languageSelector: vscode.DocumentFilter = { language, scheme: 'file' };
+	const settings = vscode.workspace.getConfiguration(EXTENSION_LANGUAGE_ID);
+	const languageSelector: vscode.DocumentFilter = { language: EXTENSION_LANGUAGE_ID, scheme: 'file' };
 
 	// test if changing specific configuration
 	let formatterEnableDidChange = false;
-	if (event && event.affectsConfiguration(language)) {
+	if (event && event.affectsConfiguration(EXTENSION_LANGUAGE_ID)) {
 		formatterEnableDidChange = settings.format.enabled !== (formatProcessor?.settings.format.enabled || false);
 
 		if (symbolProcessor) {
```

**Comment by tomaz** [comment]
wow, thanks, this works! not sure if I could figure it out any time soon 😅

**Comment by tomaz** [comment]
Allright, now I'm done - autocompletion is also behaving much nicer:

- first argument can be accepted with comma, tab or enter, second with space, tab or enter
- newline is no longer injected into the editor after intellisense confirmation

Let me know if you'd like some changes and thanks for bearing with me...

**Comment by mborik** [comment]
I'm testing it and playing with it now and after your changes IT ROCK! 🥇 
It works so fluent and reliable that I'm keeping formatOnType enabled! 😉

**Comment by tomaz** [comment]
🤘 glad you like it! Maybe it will need some tweaking if edge cases are discovered, but generally it does feel like a small magic... Thanks again for all the help and pointers (and creating this extension in the first place of course), I'd not be able to do it without that...

**Comment by mborik** [comment]
Hi @tomaz, please test the extension after my latest fixes (in a790f93 and 24285b2), because I found that your change broke behaviour without `formatOnType` a bit - after comma was pressed it broke the line in a strange way. 🙂 
Not a big deal and It's possible that the issue has been there for ages, I dunno. 😁

**Comment by tomaz** [comment]
still works!

**Pull request merged**
